ssh/ssh_tunnel.py

```diff
@@ -0,0 +1,139 @@
+"""
+Module for creating persistent SSH connections for use with synchronous
+commands. Typically, tunnels should be invoked with a context manager to
+ensure proper cleanup. E.G.:
+with contextlib.closing(SSHTunnel(*args, **kwargs)) as tunnel:
+    tunnel.write_to_remote('/usr/local/usrpath/testfile.txt', 'test_file.txt')
+    tunnel.remote_cmd(['cat', 'test_file.txt'])
+"""
+import logging
+from contextlib import closing
+from subprocess import (CalledProcessError, TimeoutExpired, check_call,
+                        check_output)
+
+logger = logging.getLogger(__name__)
+
+SOCKET_DIR = '$HOME/.cache/ssh'
+SSH_CMD = [
+    '/usr/bin/ssh',
+    '-oConnectTimeout=10',
+    '-oControlMaster=auto',
+    '-oControlPath={}/%r@%h:%p'.format(SOCKET_DIR),
+    '-oStrictHostKeyChecking=no',
+    '-oUserKnownHostsFile=/dev/null',
+    '-oBatchMode=yes',
+    '-oPasswordAuthentication=no']
+
+
+class SSHTunnel():
+
+    def __init__(self, ssh_user, ssh_key_path, host, port=22):
+        """Persistent SSH tunnel to avoid re-creating the same connection
+        Note: this should always be instantiated with contextlib.closing
+            e.g.: "with closing(SSHTunnel(*args, **kwargs)) as tunnel:"
+
+        Args:
+            ssh_user: (str) user with access to host
+            ssh_key_path: (str) local path w/ permissions to ssh_user@host
+            host: (str) locally resolvable hostname to tunnel to
+            port: (int) port to connect to host via
+
+        Return:
+            established SSHTunnel that can be issued copy/cmd/close
+        """
+        self.host = host
+        self.ssh_user = ssh_user
+        self.ssh_key_path = ssh_key_path
+        self.target = ssh_user + '@' + host
+        self.port = port
+        start_tunnel = SSH_CMD + [
+            '-fnN',
+            '-i',  ssh_key_path,
+            '-p', str(port), self.target]
+        start_tunnel = ' '.join(start_tunnel)
+        logger.debug('Starting SSH tunnel: ' + start_tunnel)
+
+        # Make sure we can cache the socket before starting
+        check_call('mkdir -p '+SOCKET_DIR, shell=True)
```

Comment:
None of this code should ever be using `shell=True`. Should always code the parameter list directly as a list of strings, in this case `['mkdir', '-p', SOCKET_DIR]`. For recursively making directories we also already have some helper functions which would be better to use